python-package/lightgbm/basic.py

```diff
@@ -2309,29 +2374,32 @@ def __init_from_csr(
         assert csr.shape[1] <= _MAX_INT32
         csr_indices = csr.indices.astype(np.int32, copy=False)
 
-        _safe_call(_LIB.LGBM_DatasetCreateFromCSR(
-            ptr_indptr,
-            ctypes.c_int(type_ptr_indptr),
-            csr_indices.ctypes.data_as(ctypes.POINTER(ctypes.c_int32)),
-            ptr_data,
-            ctypes.c_int(type_ptr_data),
-            ctypes.c_int64(len(csr.indptr)),
-            ctypes.c_int64(len(csr.data)),
-            ctypes.c_int64(csr.shape[1]),
-            _c_str(params_str),
-            ref_dataset,
-            ctypes.byref(self._handle)))
+        _safe_call(
+            _LIB.LGBM_DatasetCreateFromCSR(
+                ptr_indptr,
+                ctypes.c_int(type_ptr_indptr),
+                csr_indices.ctypes.data_as(ctypes.POINTER(ctypes.c_int32)),
+                ptr_data,
+                ctypes.c_int(type_ptr_data),
+                ctypes.c_int64(len(csr.indptr)),
+                ctypes.c_int64(len(csr.data)),
+                ctypes.c_int64(csr.shape[1]),
+                _c_str(params_str),
+                ref_dataset,
+                ctypes.byref(self._handle),
+            )
+        )
         return self
 
     def __init_from_csc(
         self,
         csc: scipy.sparse.csc_matrix,
         params_str: str,
-        ref_dataset: Optional[_DatasetHandle]
+        ref_dataset: Optional[_DatasetHandle],
```

Comment:
Commenting on this part of the diff to share a general point.... I added many of the trailing commas in this PR manually.

Without that trailing command, `ruff` / `black` wants to turn this:

```python
def __init_from_csc(
    self,
    csc: scipy.sparse.csc_matrix,
    params_str: str,
    ref_dataset: Optional[_DatasetHandle]
) -> "Dataset"
```

Into this:

```python
def __init_from_csc(
    self, csc: scipy.sparse.csc_matrix, params_str: str, ref_dataset: Optional[_DatasetHandle]
) -> "Dataset"
```

For details on that, see https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#the-magic-trailing-comma.

And this:

https://github.com/microsoft/LightGBM/blob/1b792e716682254c33ddb5eb845357e84018636d/python-package/pyproject.toml#L121